### PR TITLE
Added utility functions for unicode strings and vmi_lock_guard

### PR DESCRIPTION
**Pull request by nystilaps**
Added:
utility functions for unicode strings
vmi_lock_guard

### CODE REVIEW

**Comment by tklengyel** [comment]
```
drakvuf.c:680:9: warning: Potential leak of memory pointed to by 'out'
        return NULL;
        ^~~~~~~~~~~
1 warning generated.

**Comment by nystilaps** [comment]
677: if (VMI_SUCCESS != rc)
678: {
679:     vmi_free_unicode_str(out); // seems like there is actually no leak 
680:     return NULL;
681: }

File: src/libdrakvuf/drakvuf.c [hunk 1]
```diff
@@ -622,3 +622,74 @@ int drakvuf_event_fd_add(drakvuf_t drakvuf, int fd, event_cb_t event_cb, void* d
...
+        g_free(us.contents);
+        return NULL;
+    }
+
+    status_t rc = vmi_convert_str_encoding(&us, out, "UTF-8");
+    g_free(us.contents);
+
+    if (VMI_SUCCESS != rc)
+    {
+        vmi_free_unicode_str(out);
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c]: Just use `g_free`

Comment (nystilaps) [reply src/libdrakvuf/drakvuf.c]: This function clean up the internals too:
void
vmi_free_unicode_str(
    unicode_string_t *p_us)
{
    if (p_us->contents)
        free(p_us->contents);
    memset((void *) p_us, 0, sizeof(*p_us));
    free(p_us);
}

Comment (tklengyel) [reply src/libdrakvuf/drakvuf.c]: But Travis is still not happy about it so..

Comment (nystilaps) [reply src/libdrakvuf/drakvuf.c]: Ok, pushed fix for Travis.

File: src/libdrakvuf/drakvuf.c [hunk 2]
```diff
@@ -622,3 +622,77 @@ int drakvuf_event_fd_add(drakvuf_t drakvuf, int fd, event_cb_t event_cb, void* d
...
+    return out;
+}
+
+unicode_string_t* read_wchar_string(vmi_instance_t vmi, const access_context_t* ctx)
+{
+    size_t strlen = wchar_string_length(vmi, ctx);
+    return read_wchar_array(vmi, ctx, strlen);
+}
+
+void safe_free_unicode_str(unicode_string_t* us)
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c]: `vmi_free_unicode_str` is already NULL-safe so why this wrapper?

File: src/libdrakvuf/libdrakvuf.h [hunk 3]
```diff
@@ -414,6 +414,11 @@ int drakvuf_event_fd_add(drakvuf_t drakvuf,
 int drakvuf_event_fd_remove(drakvuf_t drakvuf,
                             int fd);
 
+unicode_string_t* read_wchar_array(vmi_instance_t vmi, const access_context_t* ctx, size_t length);
```
Comment (tklengyel) [inline src/libdrakvuf/libdrakvuf.h]: These functions will need to be prefixed with `drakvuf` like all the other functions in this header. Also, comment them to explain what they are for

File: src/libdrakvuf/drakvuf.c (hunk 2 above)
Comment (nystilaps) [reply src/libdrakvuf/drakvuf.c]: It seems like it isn't:
{
    if (p_us->contents) // <- Seems like NULL pointer dereference. Is it safe?
        free(p_us->contents);

File: src/libdrakvuf/libdrakvuf.h (hunk 3 above)
Comment (nystilaps) [reply src/libdrakvuf/libdrakvuf.h]: Ok

File: src/libdrakvuf/drakvuf.c (hunk 2 above)
Comment (tklengyel) [reply src/libdrakvuf/drakvuf.c]:
```
void
vmi_free_unicode_str(
    unicode_string_t *p_us)
{
    if (!p_us)
        return;
```

Comment (tklengyel) [reply src/libdrakvuf/drakvuf.c]: Are you using an old(er) LibVMI?

File: src/libdrakvuf/libdrakvuf.h (hunk 3 above)
Comment (nystilaps) [reply src/libdrakvuf/libdrakvuf.h]: Done

File: src/libdrakvuf/drakvuf.c (hunk 2 above)
Comment (nystilaps) [reply src/libdrakvuf/drakvuf.c]: The one that is pulled with command:
> git submodule update --init --recursive

File: src/libdrakvuf/drakvuf.c [hunk 4]
```diff
@@ -622,3 +622,90 @@ int drakvuf_event_fd_add(drakvuf_t drakvuf, int fd, event_cb_t event_cb, void* d
...
+    }
+end:
+    return (str_len / 2);
+}
+
+// Creates unicode_string_t from a string, represented like a null-terminated array of wchar (2 byte symbols)
+// In 1: access_context_t* 'ctx' with address of string head
+// In 2: string 'length' as size_t
+// Out: unicode string as unicode_string_t*; NULL on error
+unicode_string_t* drakvuf_read_wchar_array(vmi_instance_t vmi, const access_context_t* ctx, size_t length)
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c]: How is this different from `vmi_read_unicode_str`?

File: src/libdrakvuf/drakvuf.c [hunk 5]
```diff
@@ -622,3 +622,90 @@ int drakvuf_event_fd_add(drakvuf_t drakvuf, int fd, event_cb_t event_cb, void* d
...
+unicode_string_t* drakvuf_read_wchar_string(vmi_instance_t vmi, const access_context_t* ctx)
+{
+    size_t strlen = drakvuf_wchar_string_length(vmi, ctx);
+    return drakvuf_read_wchar_array(vmi, ctx, strlen);
+}
+
+// Safely free memory for any unicode_string_t*
+// In: unicode string as unicode_string_t*
+// Out: void
+void drakvuf_safe_free_unicode_str(unicode_string_t* us)
```
Comment (tklengyel) [inline src/libdrakvuf/drakvuf.c]: This function is definitely not needed, as I pointed out `vmi_free_unicode_str` is already NULL-safe.

Comment (tklengyel) [reply src/libdrakvuf/drakvuf.c]: I see that the LibVMI submodule needs an update to have that safety added, see #416

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test.

File: src/libdrakvuf/drakvuf.c (hunk 5 above)
Comment (nystilaps) [reply src/libdrakvuf/drakvuf.c]: Removed the function

**Comment by nystilaps** [comment]
Are there any other changes needed?

**Comment by tklengyel** [comment]
Well, the PR has to get through Travis. I'm also not sure whether `read_wchar_string` is really needed, how is it different from `vmi_read_unicode_str`?

**Comment by nystilaps** [comment]
Would you please possibly share your approach towards wchar arrays?

пт, 15 июн. 2018 г., 23:24 Tamas K Lengyel <notifications@github.com>:

**Comment by tklengyel** [comment]
Look at https://github.com/libvmi/libvmi/blob/master/libvmi/os/windows/unicode.c, it looks very much like this code. I would you to explain how its difreren, "i guess" doesn't convince me.

**Comment by nystilaps** [comment]
Windows has at least two different representation of strings: a unicode
string struct and a null-terminated array of UTF_16 characters.
The function you linked operates on
Unicode string struct, that has several fields, including a pointer to
string bytes. These bytes may represent ASCII, UTF_8 or UTF_16 string,
depending on string type information in that struct.
My function operates on different string representation, an null terminated
array of UTF_16 characters (words).
These functions are not interchangeable because of different domain.

сб, 16 июн. 2018 г., 4:08 Tamas K Lengyel <notifications@github.com>:

**Comment by tklengyel** [comment]
Thanks, it makes more sense now. That description should be in the commit message. Once the PR passes Travis I'll merge.

**Comment by nystilaps** [comment]
Could you point out what is the problem, found by Travis?
I cannot reproduce compilation problem locally.

**Comment by tklengyel** [comment]
You can see yourself at https://travis-ci.org/tklengyel/drakvuf/jobs/394529754

**Comment by tklengyel** [comment]
```
drakvuf.c:681:9: warning: Potential leak of memory pointed to by 'out'
        return NULL;
        ^~~~~~~~~~~
1 warning generated.

**Comment by nystilaps** [comment]
Thank you, I totally missed it.

**Comment by nystilaps** [comment]
Fixed.

File: src/libdrakvuf/libdrakvuf.h (hunk 3 above)
Comment (tklengyel) [reply src/libdrakvuf/libdrakvuf.h]: The functions stilldon't seem to be properly prefixed here.

**Comment by nystilaps** [comment]
Added prefix and comments.

**Pull request merged**